cy.contains() improvements: normalizing whitespaces, case-sensi…

sainthkh opened the pull request:
- Closes #92 
- Closes #2785

These 2 are fixed together because the implementation for one can affect the other. 

### User facing changelog

* When there are multiple white spaces, it is replaced to a single white space except in `<pre>` tag. 
* Added `matchCase` option to `cy.contains()`. 

### Additional details

#### Why was this change necessary?

##### issue #92
> Sometimes when I issue .contains() commands with text I expect to be found, it doesn't match my element as expected due to extra whitespace/new lines in my element.
> 
> **HTML**
>
>
> ```html
> <h1>hello
> world</h1>
> ```
> **Test**
> 
> ```js
> cy.contains("hello world") 
> // CypressError: Timed out retrying: Expected to find content: 'hello world' but never did.
> ```

##### issue #2785

We can currently use regex, but in some cases, we need Cypress's out-of-the-box option. 

#### What is affected by this change?

`cy.contains()` can do more things like ignoring multiple spaces and case-sensitivity.

#### Any implementation details to explain?

A new `contains` implementation has been added. 

### How has the user experience changed?

![Screenshot from 2019-11-11 17-54-59](https://user-images.githubusercontent.com/8130013/68574117-8a560280-04ac-11ea-83d2-24a56e5d5d93.png)

### PR Tasks

- [x] Have tests been added/updated?
- [x] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? => cypress-io/cypress-documentation#2397
- [x] Have API changes been updated in the [`type definitions`](cli/types/index.d.ts)?~~
- ~~Have new configuration options been added to the [`cypress.schema.json`](cli/schema/cypress.schema.json)?~~

Documentation and type definition tasks will be started when this PR is ready and reviewed.

chrisbreiding commented:
Is it possible this is a breaking change? Could a user have tests that rely on the old behavior and they will fail once this is introduced? Just need to know so that we can hold off on releasing this until the next major version of Cypress.

sainthkh commented:
@chrisbreiding #2785 doesn't break anything but #92 can be a breaking change. Then, should I change the target branch to `v4.0-release`?

jennifer-shehane commented:
@sainthkh Yes, breaking changes should target `v4.0-release` release

sainthkh commented:
Rebased and tests all passed.

chrisbreiding commented:
I think the only thing left for this is to update the documentation. @sainthkh Can you open a PR targeting https://github.com/cypress-io/cypress-documentation/pull/1710? Please update the `cy.contains()` page to detail this new behavior and update the 4.0 migration guide with an example of a test that will break with this change and how to fix it.

sainthkh commented:
1. Added option type for `cy.contains()`.
2. Wrote documentation at cypress-io/cypress-documentation#2397
3. Added a new test for leading/trailing whitespaces.

jennifer-shehane commented on packages/driver/src/cy/commands/querying.js:
```diff
@@ -477,10 +479,42 @@ module.exports = (Commands, Cypress, cy) => {
+      const normalizeWhitespaces = (elem) => {
+        let testText = elem.textContent || elem.innerText || $.text(elem)
+
+        if (elem.tagName === 'PRE') {
```
why the exception for `pre` tags? I mean, I can kind of get the logic, but it does seem strange to make an exception for a single tag and this not be documented anywhere. I guess that's my concern. Someone opening an issue one day like, 'this is working differently here' and we have this undocumented exception.

sainthkh replied:
`pre` tag preserves whitespaces like this:

https://codepen.io/sainthkh/pen/jOEpzmX

So, it is the exception. Because I thought it reflects how browsers work.

jennifer-shehane replied:
Yeah...probably adding a section in the Notes of the `.contains()` doc so that people can find this somehow. There's are other exception notes in there.

mariusnn replied:
What about elements styled with `white-space: pre;`?
We have cases where upgrading to Cypress 4 breaks tests where we're explicitily want to vaildate that the whitespace is present as it is rendered. Should there be a flag allowing override of this behaviour?

sainthkh replied:
@mariusnn It's going on #6411.

sainthkh commented:
```js
// <h1>hello world</h1>
cy.contains(/Hello World/i, { matchCase: true }) 
```

In this case, how about throwing an error:

```
cy.contains() option `matchCase` and regex flag conflicts. 
```

sainthkh commented:
Now, it throws an error for `cy.contains(/Hello World/i, { matchCase: true }) `

chrisbreiding commented on packages/driver/src/cy/commands/querying.js:
```diff
@@ -407,7 +409,11 @@ module.exports = (Commands, Cypress, cy) => {
 
-      _.defaults(options, { log: true })
+      if (options.matchCase === true && _.isRegExp(text) && text.flags.includes('i')) {
+        throw new Error('cy.contains() content has i flag and matchCase is true. What is intended?')
```
Let's move this into `error_messages.coffee` and use `$utils.throwErrByPath`. The goal is to have all error messages in there. Also, let's change the text to "You passed a regular expression with the case-insensitive (i) flag and `{ matchCase: true }` to cy.contains(). Those options conflict with each other, so please choose one or the other.`

sainthkh commented:
Fixed how error message is thrown. But it fails a lot because of "Error: Resolution method is overspecified. Specify a callback *or* return a Promise; not both."

It was introduced in bcc48f226e730cf1624f3d6379faaa887e22f16f. Needs to find way to fix it.

jennifer-shehane commented:
Yah, the 4.0 branch is broken atm, so trying to fix that up to get these tests passing.
